File: sdks/typescript/src/http-client.ts
export class HTTPError extends Error {
  status: number;
  body: Uint8Array;
  constructor(status: number, body: Uint8Array) {
    super(`http status ${status}`);
    this.name = "HTTPError";
    this.status = status;
    this.body = body;
  }
}

export class HttpClient {
  private timeoutMs: number;

  constructor(timeoutMs = 30_000) {
    this.timeoutMs = timeoutMs;
  }

  async do<T>(
    method: string,
    url: string,
    headers: Record<string, string>,
    body?: unknown,
  ): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const init: RequestInit = {
        method,
        headers: { ...headers },
        signal: controller.signal,
      };
      if (body !== undefined) {
        (init.headers as Record<string, string>)["Content-Type"] =
          "application/json";
        init.body = JSON.stringify(body);
      }
      const resp = await fetch(url, init);
      const raw = new Uint8Array(await resp.arrayBuffer());
      if (!resp.ok) {
        throw new HTTPError(resp.status, raw);
      }
      if (raw.length === 0) return undefined as T;
      return JSON.parse(new TextDecoder().decode(raw)) as T;
    } finally {
      clearTimeout(timer);
    }
  }

  async doRaw(
    method: string,
    url: string,
    headers: Record<string, string>,
    body?: BodyInit,
    contentType?: string,
  ): Promise<{ status: number; body: Uint8Array }> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const h: Record<string, string> = { ...headers };
      if (contentType) h["Content-Type"] = contentType;
      const resp = await fetch(url, {
        method,
        headers: h,
        body,
        signal: controller.signal,
      });
      const raw = new Uint8Array(await resp.arrayBuffer());
      if (!resp.ok) {
        throw new HTTPError(resp.status, raw);
      }
      return { status: resp.status, body: raw };
    } finally {
      clearTimeout(timer);
    }
  }

  async fetchRaw(url: string, init: RequestInit): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(url, { ...init, signal: controller.signal });
      if (!resp.ok) {
        const raw = new Uint8Array(await resp.arrayBuffer());
        throw new HTTPError(resp.status, raw);
      }
      clearTimeout(timer); // Don't timeout during streaming
      return resp;
    } catch (e) {
      clearTimeout(timer);
      throw e;
    }
  }
}

File: sdks/python/src/sandboxer/_http_client.py
from __future__ import annotations

import httpx


class HTTPError(Exception):
    def __init__(self, status: int, body: bytes):
        super().__init__(f"http status {status}")
        self.status = status
        self.body = body


class HttpClient:
    def __init__(self, timeout_s: float = 30.0):
        self._client = httpx.Client(timeout=timeout_s)

    def do(self, method: str, url: str, headers: dict[str, str], body: dict | None = None) -> dict | list | None:
        kwargs: dict = {"method": method, "url": url, "headers": headers}
        if body is not None:
            kwargs["json"] = body
        resp = self._client.request(**kwargs)
        if resp.status_code >= 400:
            raise HTTPError(resp.status_code, resp.content)
        if not resp.content:
            return None
        return resp.json()

    def do_raw(
        self,
        method: str,
        url: str,
        headers: dict[str, str],
        content: bytes | None = None,
        content_type: str | None = None,
    ) -> tuple[int, bytes]:
        h = dict(headers)
        if content_type:
            h["Content-Type"] = content_type
        resp = self._client.request(method=method, url=url, headers=h, content=content)
        if resp.status_code >= 400:
            raise HTTPError(resp.status_code, resp.content)
        return resp.status_code, resp.content

    def request(self, method: str, url: str, headers: dict[str, str], **kwargs) -> httpx.Response:
        """Low-level request for multipart uploads etc."""
        return self._client.request(method=method, url=url, headers=headers, **kwargs)

    def close(self) -> None:
        self._client.close()


class AsyncHttpClient:
    def __init__(self, timeout_s: float = 30.0):
        self._client = httpx.AsyncClient(timeout=timeout_s)

    async def do(self, method: str, url: str, headers: dict[str, str], body: dict | None = None) -> dict | list | None:
        kwargs: dict = {"method": method, "url": url, "headers": headers}
        if body is not None:
            kwargs["json"] = body
        resp = await self._client.request(**kwargs)
        if resp.status_code >= 400:
            raise HTTPError(resp.status_code, resp.content)
        if not resp.content:
            return None
        return resp.json()

    async def do_raw(
        self,
        method: str,
        url: str,
        headers: dict[str, str],
        content: bytes | None = None,
        content_type: str | None = None,
    ) -> tuple[int, bytes]:
        h = dict(headers)
        if content_type:
            h["Content-Type"] = content_type
        resp = await self._client.request(method=method, url=url, headers=h, content=content)
        if resp.status_code >= 400:
            raise HTTPError(resp.status_code, resp.content)
        return resp.status_code, resp.content

    async def request(self, method: str, url: str, headers: dict[str, str], **kwargs) -> httpx.Response:
        return await self._client.request(method=method, url=url, headers=headers, **kwargs)

    async def close(self) -> None:
        await self._client.aclose()
